### PR TITLE
fix(minor-rewards)!: handle signing rewards for different chains

### PR DESCRIPTION
**Pull request by haiyizxx**
## Description
The rewards contract can not distinguish signing events for different chains. All signing events are added to a single tally, since all signing events are handled by one multisig contract.

This pr associates rewards with a (chain name, contract address) pair.

## Todos

- [x] Unit tests
- [x] Manual tests
- [x] Documentation
- [x] Connect epics/issues

### CODE REVIEW

File: contracts/rewards/src/state.rs [hunk 1]
```diff
@@ -168,18 +241,18 @@ pub trait Store {
...
 
-/// Maps a (contract address, epoch number) pair to a tally for that epoch and contract
-const TALLIES: Map<(Addr, u64), EpochTally> = Map::new("tallies");
+/// Maps a (pool id, epoch number) pair to a tally for that epoch and rewards pool
+const TALLIES: Map<TallyKey, EpochTally> = Map::new("tallies");
 
-/// Maps an (event id, contract address) pair to an Event
-const EVENTS: Map<(String, Addr), Event> = Map::new("events");
+/// Maps an (event id, pool id) pair to an Event
+const EVENTS: Map<(String, PoolId), Event> = Map::new("events");
```
Comment (cjcobb23) [inline contracts/rewards/src/state.rs:248]: might be worthwhile to use a type alias for String here, to make things clearer.

File: contracts/rewards/src/state.rs [hunk 2]
```diff
@@ -192,35 +265,39 @@ impl Store for RewardsStore<'_> {
...
         epoch_num: u64,
     ) -> Result<Option<EpochTally>, ContractError> {
         TALLIES
-            .may_load(self.storage, (contract, epoch_num))
+            .may_load(self.storage, TallyKey { pool_id, epoch_num })
             .change_context(ContractError::LoadEpochTally)
     }
 
-    fn load_rewards_pool(&self, contract: Addr) -> Result<RewardsPool, ContractError> {
+    fn load_rewards_pool(&self, id: PoolId) -> Result<RewardsPool, ContractError> {
```
Comment (cjcobb23) [inline contracts/rewards/src/state.rs:294]:
```suggestion
    fn load_rewards_pool(&self, pool_id: PoolId) -> Result<RewardsPool, ContractError> {
```

**Review by cjcobb23 — APPROVED** [review]

File: contracts/rewards/src/msg.rs [hunk 3]
```diff
@@ -35,12 +36,14 @@ pub enum ExecuteMsg {
...
+        chain_name: ChainName,
         event_id: nonempty::String,
         worker_address: String,
     },
 
     /// Distribute rewards up to epoch T - 2 (i.e. if we are currently in epoch 10, distribute all undistributed rewards for epochs 0-8) and send the required number of tokens to each worker
     DistributeRewards {
+        chain_name: ChainName,
         /// Address of contract for which to process rewards. For example, address of a voting verifier instance.
         contract_address: String,
```
Comment (fish-sammy) [inline contracts/rewards/src/msg.rs:48]:
```suggestion
        pool_id: PoolId,
```

File: contracts/rewards/src/msg.rs [hunk 4]
```diff
@@ -50,6 +53,7 @@ pub enum ExecuteMsg {
     /// Start a new reward pool for the given contract if none exists. Otherwise, add tokens to an existing reward pool.
     /// Any attached funds with a denom matching the rewards denom are added to the pool.
     AddRewards {
+        chain_name: ChainName,
         /// Address of contract for which to reward participation. For example, address of a voting verifier instance.
         contract_address: String,
```
Comment (fish-sammy) [inline contracts/rewards/src/msg.rs:58]:
```suggestion
        pool_id: PoolId,
```

File: contracts/rewards/src/state.rs [hunk 5]
```diff
@@ -23,19 +24,87 @@ pub struct StoredParams {
...
+
+impl<'a> Prefixer<'a> for PoolId {
+    fn prefix(&self) -> Vec<Key> {
+        self.key()
+    }
+}
+
+#[cw_serde]
+#[derive(Eq, Hash)]
+pub struct TallyKey {
```
Comment (fish-sammy) [inline contracts/rewards/src/state.rs:68]:
```suggestion
pub struct TallyId {
```

File: contracts/rewards/src/state.rs (hunk 1 above)
Comment (haiyizxx) [reply contracts/rewards/src/state.rs:248]: Agree, can do in another pr. There some other places using event id.

**Pull request merged**